### ISSUE TITLE
unclear behavior of keymap function callback for visual mode mappings

### ISSUE THREAD

**Issue by rebelot** [neovim/neovim#18340]
### Neovim version (nvim -v)

v0.8.0-dev+96-gd9dcfd021

### Vim (not Nvim) behaves the same?

-

### Operating system/version

macOS 12.2

### Terminal name/version

kitty 0.25.0

### $TERM environment variable

tmux-256color

### Installation

build

### How to reproduce the issue

```lua
function Print_range()
    local pos = {
        start = vim.fn.getpos("'<"),
        ["end"] = vim.fn.getpos("'>"),
    }
    vim.pretty_print(pos)
end

vim.keymap.set({ "n", "x" }, "<leader>v", Print_range)
vim.keymap.set({ "n", "x" }, "<leader>V", ":lua Print_range()<cr>")
```

enter visual mode on line 1 and use `<leader>v`
enter visual mode on line 2 and use `<leader>V`

### Expected behavior

the two mappings behave identically in visual mode.

### Actual behavior

when using the callback mapping during visual mode, visual range position is not updated.

it would be nice to control the execution of the callback as if it was executed using `":<C-u>call luaref()<cr>"` or `"<cmd>call luaref()<cr>"`

**Comment by clason** [comment]
Yes, because `:` enters command mode, so these two mappings are in fact _not_ equivalent.

**Comment by rebelot** [comment]
@clason 
>Yes, because : enters command mode (as opposed to <cmd>, see :h <Cmd>), so these two mappings are in fact not equivalent.

I know, but you can in fact control the behavior when using "old-style" mappings. This is not controllable when using callbacks.

This invalidates visual-mode mappings that leverage callbacks and closures. The point of visual-mode mappings is indeed to _know_ where visual selection starts/ends

**Comment by clason** [comment]
The point is that the right fix here is to allow more direct access to this information from Lua (e.g., via something to https://github.com/neovim/neovim/pull/13896) rather than adding the "magic" Vim behavior to the Neovim API.

**Comment by rebelot** [comment]
I'm happy as long as it's a WIP! 

btw, I'd just mention it somewhere that lsp range code actions mapping _will not_ work if using something like

```lua
    vim.keymap.set(
        "x",
        "<leader>la",
        vim.lsp.buf.range_code_action
    }
```
